grails-app/services/org/pih/warehouse/data/ProductSupplierAttributeDataService.groovy

```diff
@@ -0,0 +1,135 @@
+/**
+ * Copyright (c) 2012 Partners In Health.  All rights reserved.
+ * The use and distribution terms for this software are covered by the
+ * Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
+ * which can be found in the file epl-v10.html at the root of this distribution.
+ * By using this software in any fashion, you are agreeing to be bound by
+ * the terms of this license.
+ * You must not remove this notice, or any other, from this software.
+ **/
+package org.pih.warehouse.data
+
+import org.pih.warehouse.core.EntityTypeCode
+import org.pih.warehouse.core.UnitOfMeasure
+import org.pih.warehouse.importer.ImportDataCommand
+import org.pih.warehouse.product.Attribute
+import org.pih.warehouse.product.Product
+import org.pih.warehouse.product.ProductAttribute
+import org.pih.warehouse.product.ProductSupplier
+
+class ProductSupplierAttributeDataService {
+
+    Boolean validate(ImportDataCommand command) {
+        log.info "Validate data " + command.filename
+        command.data.eachWithIndex { params, index ->
+
+            def productCode = params.productCode
+            def productSupplierCode = params.productSupplierCode
+            def attributeCode = params.attributeCode
+            def attributeValue = params.attributeValue
+            def unitOfMeasure = params.unitOfMeasure
+
+            // Product code is required
+            if (!productCode) {
+                command.errors.reject("Row ${index + 1}: Product code is required")
+                return
+            }
+
+            // Product should exist
+            Product product = Product.findByProductCode(productCode)
+            if (!product) {
+                command.errors.reject("Row ${index + 1}: Product with code: ${productCode} does not exist")
+                return
+            }
+
+            // Product supplier code is required
+            if (!productSupplierCode) {
+                command.errors.reject("Row ${index + 1}: Product source code is required")
+                return
+            }
+
+            // Product supplier should exist
+            ProductSupplier productSupplier = ProductSupplier.findByCode(productSupplierCode)
+            if (!productSupplier) {
+                command.errors.reject("Row ${index + 1}: Product source with code: ${productSupplierCode} does not exist")
+                return
+            }
+
+            // Attribute code is required
+            if (!attributeCode) {
+                command.errors.reject("Row ${index + 1}: Attribute code is required")
+                return
+            }
+
+            // Attribute should exist
+            Attribute attribute = Attribute.findByCode(attributeCode)
+            if (!attribute) {
+                command.errors.reject("Row ${index + 1}: Attribute with code: ${attributeCode} does not exist")
```

Comment:
remove the colon after code

Reply:
i'm going to stop adding these comments, please just find all "code:" and replace with "code"